slsim/Halos/halos.py

```diff
@@ -0,0 +1,758 @@
+    a statistical basis for modeling distributions of astronomical objects over
+    redshift.
+
+    :param N: The number of redshifts to sample from the cumulative distribution,
+        representing the size of the resultant redshift array.
+    :param dN_dz: Differential number of objects per redshift interval, indicating the
+        rate at which the number of objects changes with redshift.
+    :param redshift_list: Array of redshifts corresponding to the dN_dz values,
+        delineating the domain over which the distribution applies.
+    :type N: int
+    :type dN_dz: ndarray,
+    :type redshift_list: ndarray,
+    :return: An array of N redshifts sampled according to the cumulative distribution
+        derived from the differential redshift distribution dN/dz.
+    :rtype: ndarray
+    """
+    assert len(dN_dz) == len(redshift_list)
+    cdf = dN_dz  # reuse memory
+    np.cumsum((dN_dz[1:] + dN_dz[:-1]) / 2 * np.diff(redshift_list), out=cdf[1:])
+    cdf[0] = 0
+    cdf /= cdf[-1]
+    return np.interp(np.random.rand(N), cdf, redshift_list)
+
+
+def v_per_redshift(redshift_list, cosmology, sky_area):
+    """Calculate comoving volume per unit redshift.
+
+    This function computes the comoving volume associated with each redshift in a given list of redshifts. The calculation considers the cosmological model specified and the area of the sky under consideration. This is crucial for understanding the volume over which astronomical surveys operate at different depths (redshifts).
+
+    :param redshift_list: Array of redshifts for which to calculate the corresponding comoving volumes, representing the depth of an astronomical survey or observation.
+    :param cosmology: The cosmological model to apply, which defines the universe's geometry and expansion history, influencing the calculation of comoving volumes. This parameter is optional, with a default cosmology used if not specified.
+    :param sky_area: The area of the sky over which the volume calculations are to be applied, expressed in square degrees or as a solid angle, framing the scope of the astronomical observation or survey.
+    :type redshift_list: array_like
+    :type cosmology: astropy.Cosmology instance, optional
+    :type sky_area: `~astropy.units.Quantity`
+    :return: An array detailing the comoving volume corresponding to each redshift in the input list, providing a spatial context for the distribution of astronomical objects.
+    :rtype: array
+    """
+    dV_dz = (cosmology.differential_comoving_volume(redshift_list) * sky_area).to_value(
+        "Mpc3"
+    )
+    return dV_dz
+
+
+def halo_mass_at_z(
+    z,
+    m_min=None,
+    m_max=None,
+    resolution=None,
+    cosmology=None,
+    sigma8=0.81,
+    ns=0.96,
+    omega_m=None,
+):
+    """Sample halo masses at given redshift(s) using specified or default cosmological
+    parameters and mass range.
+
+    :param z: Redshift(s) at which to sample halo masses. Can be a single value or an
```

Comment:
I am not so this is clear from the documentation. These are the sets of redshifts, each corresponding to one halo that needs to be drawn from the hmf, correct?

Reply:
this function is actually not used now since colossus just read the redshift